File: codes/backtraking/90.subsets-ii/subsets-ii.cpp
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
    using vit = vector<int>::const_iterator;
    vector<int> path;
    vector<vector<int>> result;

public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        result.push_back(vector<int>{});
        backtracing(nums.cbegin(), nums.cend());
        return result;
    }

    void backtracing(vit beg, vit end) {
        if (beg == end) {
            return;
        }
        for (auto it = beg; it < end; it++) {
            if (it != beg && *(it - 1) == *it) {
                continue;
            }
            path.push_back(*it);
            result.push_back(path);
            backtracing(it + 1, end);
            path.pop_back();
        }
    }
};

File: codes/backtraking/90.subsets-ii/subsets-ii.go
package main

import "sort"

func subsetsWithDup(nums []int) [][]int {
	result := [][]int{[]int{}}
	path := []int{}
	var backtracing func(int)
	backtracing = func(beg int) {
		if beg == len(nums) {
			return
		}
		for i := beg; i < len(nums); i++ {
			if i == beg && nums[i-1] == nums[i] {
				continue
			}
			path = append(path, nums[i])
			result = append(result, append([]int{}, path...))
			backtracing(i + 1)
			path = path[:len(path)-1]
		}
	}

	sort.Ints(nums)
	backtracing(0)
	return result

}
